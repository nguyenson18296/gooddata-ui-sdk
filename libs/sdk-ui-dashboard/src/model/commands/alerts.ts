// (C) 2021 GoodData Corporation

import { IWidgetAlert, IWidgetAlertDefinition } from "@gooddata/sdk-backend-spi";
import { IDashboardCommand } from "./base";

/**
 * Creates Kpi alert.
 *
 * @alpha
 */
export interface CreateAlert extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.ALERT.CREATE";
    readonly payload: {
        readonly alert: IWidgetAlertDefinition;
    };
}

/**
 * Creates the SaveAlert command. Dispatching this command will result in the creating Kpi alert on the backend.
 *
 * @param alert - specify alert to create.
 * @param correlationId - optionally specify correlation id to use for this command. this will be included in all
 *  events that will be emitted during the command processing
 * @alpha
 */
export function createAlert(alert: IWidgetAlertDefinition, correlationId?: string): CreateAlert {
    return {
        type: "GDC.DASH/CMD.ALERT.CREATE",
        correlationId,
        payload: {
            alert,
        },
    };
}

/**
 *Updates Kpi alert.
 *
 * @alpha
 */
export interface UpdateAlert extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.ALERT.UPDATE";
    readonly payload: {
        readonly alert: IWidgetAlert;
    };
}

/**
 * Creates the UpdateAlert command. Dispatching this command will result in the updating Kpi alert on the backend.
 *
 * @param alert - specify alert to update.
 * @param correlationId - optionally specify correlation id to use for this command. this will be included in all
 *  events that will be emitted during the command processing

 * @alpha
 */
export function updateAlert(alert: IWidgetAlert, correlationId?: string): UpdateAlert {
    return {
        type: "GDC.DASH/CMD.ALERT.UPDATE",
        correlationId,
        payload: {
            alert,
        },
    };
}

//
//
//

/**
 * Removes Kpi alert.
 *
 * @alpha
 */
export interface RemoveAlert extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.ALERT.REMOVE";
    readonly payload: {
        readonly alert: IWidgetAlert;
    };
}

/**
 * Creates the RemoveAlert command. Dispatching this command will result in the removing Kpi alert on the backend.
 *
 * @param alertRef - specify ObjRef of the alert to remove
 * @param correlationId - optionally specify correlation id to use for this command. this will be included in all
 *  events that will be emitted during the command processing

 * @alpha
 */
export function removeAlert(alert: IWidgetAlert, correlationId?: string): RemoveAlert {
    return {
        type: "GDC.DASH/CMD.ALERT.REMOVE",
        correlationId,
        payload: {
            alert,
        },
    };
}

//
//
//
