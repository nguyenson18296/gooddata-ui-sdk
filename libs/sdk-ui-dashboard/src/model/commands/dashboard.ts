// (C) 2021 GoodData Corporation

import { DashboardConfig } from "../types/commonTypes";
import { IWorkspacePermissions } from "@gooddata/sdk-backend-spi";
import { IDashboardCommand } from "./base";

/**
 * The initial load of the dashboard will use this correlation id.
 *
 * @alpha
 */
export const InitialLoadCorrelationId = "initialLoad";

/**
 * Loads dashboard from analytical backend.
 *
 * @alpha
 */
export interface LoadDashboard extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.LOAD";
    readonly payload: {
        readonly config?: DashboardConfig;
        readonly permissions?: IWorkspacePermissions;
    };
}

/**
 * Creates the LoadDashboard command. Dispatching this command will result in the load of all
 * the essential data from the backend and initializing the state of Dashboard to a point where the
 * dashboard can be rendered.
 *
 * @param config - optionally specify configuration to use for for the Dashboard; you MAY provide partial configuration.
 *  During the LoadDashboard processing the Dashboard component will resolve all the missing parts by reading them
 *  from the backend.
 * @param permissions - optionally specify permissions to use when determining whether the user is eligible for some
 *  actions with the dashboard; if you do not specify permissions Dashboard component will load the permissions
 *  from the backend.
 * @param correlationId - optionally specify correlation id to use for this command. this will be included in all
 *  events that will be emitted during the command processing
 *
 * @alpha
 */
export function loadDashboard(
    config?: DashboardConfig,
    permissions?: IWorkspacePermissions,
    correlationId?: string,
): LoadDashboard {
    return {
        type: "GDC.DASH/CMD.LOAD",
        correlationId,
        payload: {
            config,
            permissions,
        },
    };
}

//
//
//

/**
 * @alpha
 */
export interface SaveDashboard extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.SAVE";
    readonly payload: {
        readonly identifier?: string;
    };
}

/**
 * Creates the SaveDashboard command. Dispatching this command will result in persisting all the accumulated
 * dashboard modification to backend.
 *
 * The command will not have any effect if dashboard is not initialized or is empty.
 *
 * @param identifier - optionally specify identifier to set for the saved dashboard. If specified, the
 *  identifier will be used only during the initial save of a new dashboard. When the SaveDashboard is called
 *  of a dashboard that already exists on a backend, then the identifier will be ignored. If no identifier
 *  is specified for the initial save, then the identifier will be generated.
 * @param correlationId - optionally specify correlation id to use for this command. this will be included in all
 *  events that will be emitted during the command processing

 * @alpha
 */
export function saveDashboard(identifier?: string, correlationId?: string): SaveDashboard {
    return {
        type: "GDC.DASH/CMD.SAVE",
        correlationId,
        payload: {
            identifier,
        },
    };
}

//
//
//

/**
 * @alpha
 */
export interface SaveDashboardAs extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.SAVEAS";
    readonly payload: {
        readonly identifier?: string;
        readonly title?: string;
    };
}

/**
 * Creates the SaveDashboardAs command. Dispatching this command will result in creation of a copy of the
 * current dashboard. The copy will reflect the current state of the dashboard including any modifications done
 * on top of the original dashboard.
 *
 * Upon success, a copy of the dashboard will be persisted on the backend. The context of the dashboard component
 * that processed the command is unchanged - it still works with the original dashboard.
 *
 * @param identifier - identifier to assign to the newly created dashboard; if not specified a random identifier
 *  will be generated
 * @param title - new title for the dashboard; if not specified, the title of original dashboard will be used
 * @param correlationId - optionally specify correlation id to use for this command. this will be included in all
 *  events that will be emitted during the command processing
 * @alpha
 */
export function saveDashboardAs(
    identifier?: string,
    title?: string,
    correlationId?: string,
): SaveDashboardAs {
    return {
        type: "GDC.DASH/CMD.SAVEAS",
        correlationId,
        payload: {
            identifier,
            title,
        },
    };
}

//
//
//

/**
 * @alpha
 */
export interface RenameDashboard extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.RENAME";
    readonly payload: {
        readonly newTitle: string;
    };
}

/**
 * Creates the RenameDashboard command. Dispatching this command will result in rename of the dashboard. The changes
 * will be done only in-memory and have to be flushed to backend using the SaveDashboard command.
 *
 * @param newTitle - new dashboard title
 * @param correlationId - optionally specify correlation id to use for this command. this will be included in all
 *  events that will be emitted during the command processing
 * @alpha
 */
export function renameDashboard(newTitle: string, correlationId?: string): RenameDashboard {
    return {
        type: "GDC.DASH/CMD.RENAME",
        correlationId,
        payload: {
            newTitle,
        },
    };
}

//
//
//

/**
 * @alpha
 */
export interface ResetDashboard extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.RESET";
}

/**
 * Creates the ResetDashboard command. Dispatching this command will result in dropping all in-memory modifications
 * of the dashboard and reverting to a state after the initial LoadCommand.
 *
 * Note: reset dashboard will not reload dashboard data from backend.
 *
 * @param correlationId - optionally specify correlation id to use for this command. this will be included in all
 *  events that will be emitted during the command processing
 *
 * @alpha
 */
export function resetDashboard(correlationId?: string): ResetDashboard {
    return {
        type: "GDC.DASH/CMD.RESET",
        correlationId,
    };
}
