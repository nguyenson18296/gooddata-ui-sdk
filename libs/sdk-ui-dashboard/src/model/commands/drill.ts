// (C) 2021 GoodData Corporation
import { IDashboardCommand } from "./base";
import { IDashboardDrillEvent, IDrillDownDefinition } from "@gooddata/sdk-ui-ext";
import {
    IDrillToAttributeUrl,
    IDrillToCustomUrl,
    IDrillToDashboard,
    IDrillToInsight,
    IDrillToLegacyDashboard,
} from "@gooddata/sdk-backend-spi";
import { DashboardDrillContext } from "../../drill/interfaces";
import { IInsight } from "@gooddata/sdk-model";

/**
 * Performs drill.
 *
 * @internal
 */
export interface Drill extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.DRILL";
    readonly payload: {
        readonly drillEvent: IDashboardDrillEvent;
        readonly drillContext: DashboardDrillContext;
    };
}

/**
 * Creates the PerformDrill command.
 * Dispatching this command will result into triggering relevant drill events.
 *
 * @internal
 */
export function drill(
    drillEvent: IDashboardDrillEvent,
    drillContext: DashboardDrillContext,
    correlationId?: string,
): Drill {
    return {
        type: "GDC.DASH/CMD.DRILL",
        correlationId,
        payload: {
            drillEvent,
            drillContext,
        },
    };
}

//
//
//

/**
 * Performs drill down.
 *
 * @internal
 */
export interface DrillDown extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.DRILL.DRILL_DOWN";
    readonly payload: {
        readonly insight: IInsight;
        readonly drillDefinition: IDrillDownDefinition;
        readonly drillEvent: IDashboardDrillEvent;
    };
}

/**
 * Creates the PerformDrillDown command.
 * Dispatching this command will result into loading additional data and then triggering relevant drill events.
 *
 * @internal
 */
export function drillDown(
    insight: IInsight,
    drillDefinition: IDrillDownDefinition,
    drillEvent: IDashboardDrillEvent,
    correlationId?: string,
): DrillDown {
    return {
        type: "GDC.DASH/CMD.DRILL.DRILL_DOWN",
        correlationId,
        payload: {
            insight,
            drillDefinition,
            drillEvent,
        },
    };
}

//
//
//

/**
 * Performs drill to insight.
 *
 * @internal
 */
export interface DrillToInsight extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.DRILL.DRILL_TO_INSIGHT";
    readonly payload: {
        readonly drillDefinition: IDrillToInsight;
        readonly drillEvent: IDashboardDrillEvent;
    };
}

/**
 * Creates the PerformDrillToInsight command.
 * Dispatching this command will result into loading additional data and then triggering relevant drill events.
 *
 * @internal
 */
export function drillToInsight(
    drillDefinition: IDrillToInsight,
    drillEvent: IDashboardDrillEvent,
    correlationId?: string,
): DrillToInsight {
    return {
        type: "GDC.DASH/CMD.DRILL.DRILL_TO_INSIGHT",
        correlationId,
        payload: {
            drillDefinition,
            drillEvent,
        },
    };
}

//
//
//

/**
 * Performs drill to dashboard.
 *
 * @internal
 */
export interface DrillToDashboard extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.DRILL.DRILL_TO_DASHBOARD";
    readonly payload: {
        readonly drillDefinition: IDrillToDashboard;
        readonly drillEvent: IDashboardDrillEvent;
    };
}

/**
 * Creates the PerformDrillToDashboard command.
 * Dispatching this command will result into loading additional data and then triggering relevant drill events.
 *
 * @internal
 */
export function drillToDashboard(
    drillDefinition: IDrillToDashboard,
    drillEvent: IDashboardDrillEvent,
    correlationId?: string,
): DrillToDashboard {
    return {
        type: "GDC.DASH/CMD.DRILL.DRILL_TO_DASHBOARD",
        correlationId,
        payload: {
            drillDefinition,
            drillEvent,
        },
    };
}

//
//
//

/**
 * Performs drill to custom url.
 *
 * @internal
 */
export interface DrillToCustomUrl extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.DRILL.DRILL_TO_CUSTOM_URL";
    readonly payload: {
        readonly drillDefinition: IDrillToCustomUrl;
        readonly drillEvent: IDashboardDrillEvent;
    };
}

/**
 * Creates the PerformDrillToCustomUrl command.
 * Dispatching this command will result into loading additional data and then triggering relevant drill events.
 *
 * @internal
 */
export function drillToCustomUrl(
    drillDefinition: IDrillToCustomUrl,
    drillEvent: IDashboardDrillEvent,
    correlationId?: string,
): DrillToCustomUrl {
    return {
        type: "GDC.DASH/CMD.DRILL.DRILL_TO_CUSTOM_URL",
        correlationId,
        payload: {
            drillDefinition,
            drillEvent,
        },
    };
}

//
//
//

/**
 * Performs drill to attribute url.
 *
 * @internal
 */
export interface DrillToAttributeUrl extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.DRILL.DRILL_TO_ATTRIBUTE_URL";
    readonly payload: {
        readonly drillDefinition: IDrillToAttributeUrl;
        readonly drillEvent: IDashboardDrillEvent;
    };
}

/**
 * Creates the PerformDrillToAttributeUrl command.
 * Dispatching this command will result into loading additional data and then triggering relevant drill events.
 *
 * @internal
 */
export function drillToAttributeUrl(
    drillDefinition: IDrillToAttributeUrl,
    drillEvent: IDashboardDrillEvent,
    correlationId?: string,
): DrillToAttributeUrl {
    return {
        type: "GDC.DASH/CMD.DRILL.DRILL_TO_ATTRIBUTE_URL",
        correlationId,
        payload: {
            drillDefinition,
            drillEvent,
        },
    };
}

//
//
//

/**
 * Performs drill to legacy dashboard.
 *
 * @internal
 */
export interface DrillToLegacyDashboard extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.DRILL.DRILL_TO_LEGACY_DASHBOARD";
    readonly payload: {
        readonly drillDefinition: IDrillToLegacyDashboard;
        readonly drillEvent: IDashboardDrillEvent;
    };
}

/**
 * Creates the PerformDrillToLegacyDashboard command.
 * Dispatching this command will result into loading additional data and then triggering relevant drill events.
 *
 * @internal
 */
export function drillToLegacyDashboard(
    drillDefinition: IDrillToLegacyDashboard,
    drillEvent: IDashboardDrillEvent,
    correlationId?: string,
): DrillToLegacyDashboard {
    return {
        type: "GDC.DASH/CMD.DRILL.DRILL_TO_LEGACY_DASHBOARD",
        correlationId,
        payload: {
            drillDefinition,
            drillEvent,
        },
    };
}
