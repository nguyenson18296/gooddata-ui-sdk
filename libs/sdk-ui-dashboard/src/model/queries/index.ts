// (C) 2021 GoodData Corporation

import { QueryDateDatasetsForInsight } from "./insights";

export { IDashboardQuery, DashboardQueryType } from "./base";
export { QueryDateDatasetsForInsight, DateDatasetsForInsight, queryDateDatasetsForInsight } from "./insights";

/**
 * @internal
 */
export type DashboardQueries = QueryDateDatasetsForInsight;
