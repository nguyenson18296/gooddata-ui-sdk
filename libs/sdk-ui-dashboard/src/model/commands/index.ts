// (C) 2021 GoodData Corporation

import { LoadDashboard, RenameDashboard, ResetDashboard, SaveDashboard, SaveDashboardAs } from "./dashboard";
import {
    AddAttributeFilter,
    ChangeAttributeFilterSelection,
    ChangeDateFilterSelection,
    MoveAttributeFilter,
    RemoveAttributeFilters,
    SetAttributeFilterParent,
} from "./filters";
import {
    ChangeInsightWidgetFilterSettings,
    ChangeInsightWidgetHeader,
    ChangeInsightWidgetInsight,
    ChangeInsightWidgetVisProperties,
    ModifyDrillsForInsightWidget,
    RefreshInsightWidget,
    RemoveDrillsForInsightWidget,
} from "./insight";
import {
    ChangeKpiWidgetComparison,
    ChangeKpiWidgetFilterSettings,
    ChangeKpiWidgetHeader,
    ChangeKpiWidgetMeasure,
    RefreshKpiWidget,
} from "./kpi";
import {
    AddLayoutSection,
    AddSectionItems,
    ChangeLayoutSectionHeader,
    MoveLayoutSection,
    MoveSectionItem,
    RemoveLayoutSection,
    RemoveSectionItem,
    ReplaceSectionItem,
    UndoLayoutChanges,
} from "./layout";
import { CreateAlert, RemoveAlert, UpdateAlert } from "./alerts";
import { CreateScheduledEmail } from "./scheduledEmail";
import {
    Drill,
    DrillDown,
    DrillToAttributeUrl,
    DrillToCustomUrl,
    DrillToDashboard,
    DrillToInsight,
    DrillToLegacyDashboard,
} from "./drill";

export { DashboardCommandType, IDashboardCommand, CommandProcessingMeta } from "./base";
export {
    InitialLoadCorrelationId,
    LoadDashboard,
    loadDashboard,
    SaveDashboardAs,
    saveDashboardAs,
    SaveDashboard,
    saveDashboard,
    RenameDashboard,
    renameDashboard,
    ResetDashboard,
    resetDashboard,
} from "./dashboard";
export {
    ChangeDateFilterSelection,
    changeDateFilterSelection,
    clearDateFilterSelection,
    AddAttributeFilter,
    addAttributeFilter,
    MoveAttributeFilter,
    moveAttributeFilter,
    RemoveAttributeFilters,
    removeAttributeFilter,
    ChangeAttributeFilterSelection,
    AttributeFilterSelectionType,
    resetAttributeFilterSelection,
    changeAttributeFilterSelection,
    SetAttributeFilterParent,
    setAttributeFilterParent,
} from "./filters";
export {
    AddLayoutSection,
    addLayoutSection,
    MoveLayoutSection,
    moveLayoutSection,
    RemoveLayoutSection,
    removeLayoutSection,
    ChangeLayoutSectionHeader,
    changeLayoutSectionHeader,
    AddSectionItems,
    addSectionItem,
    ReplaceSectionItem,
    replaceSectionItem,
    MoveSectionItem,
    moveSectionItem,
    RemoveSectionItem,
    eagerRemoveSectionItem,
    UndoLayoutChanges,
    undoLayoutChanges,
    revertLastLayoutChange,
    DashboardLayoutCommands,
    UndoPointSelector,
} from "./layout";
export { CreateAlert, createAlert, RemoveAlert, removeAlert, UpdateAlert, updateAlert } from "./alerts";
export { CreateScheduledEmail } from "./scheduledEmail";
export {
    Drill,
    DrillDown,
    DrillToAttributeUrl,
    DrillToCustomUrl,
    DrillToDashboard,
    DrillToInsight,
    DrillToLegacyDashboard,
    drill,
    drillDown,
    drillToAttributeUrl,
    drillToCustomUrl,
    drillToDashboard,
    drillToInsight,
    drillToLegacyDashboard,
} from "./drill";
export {
    ChangeKpiWidgetHeader,
    changeKpiWidgetHeader,
    ChangeKpiWidgetMeasure,
    changeKpiWidgetMeasure,
    ChangeKpiWidgetFilterSettings,
    changeKpiWidgetFilterSettings,
    ChangeKpiWidgetComparison,
    changeKpiWidgetComparison,
    RefreshKpiWidget,
    refreshKpiWidget,
    KpiWidgetComparison,
} from "./kpi";

export {
    ChangeInsightWidgetHeader,
    changeInsightWidgetHeader,
    ChangeInsightWidgetFilterSettings,
    changeInsightWidgetFilterSettings,
    ChangeInsightWidgetVisProperties,
    changeInsightWidgetVisProperties,
    ChangeInsightWidgetInsight,
    changeInsightWidgetInsight,
    ModifyDrillsForInsightWidget,
    modifyDrillForInsightWidget,
    RemoveDrillsForInsightWidget,
    removeDrillForInsightWidget,
    RemoveDrillsSelector,
    RefreshInsightWidget,
    refreshInsightWidget,
} from "./insight";

/**
 * @internal
 */
export type DashboardCommands =
    | LoadDashboard
    | SaveDashboard
    | SaveDashboardAs
    | RenameDashboard
    | ResetDashboard
    | ChangeDateFilterSelection
    | AddAttributeFilter
    | RemoveAttributeFilters
    | MoveAttributeFilter
    | ChangeAttributeFilterSelection
    | SetAttributeFilterParent
    | AddLayoutSection
    | MoveLayoutSection
    | RemoveLayoutSection
    | ChangeLayoutSectionHeader
    | AddSectionItems
    | ReplaceSectionItem
    | MoveSectionItem
    | RemoveSectionItem
    | UndoLayoutChanges
    | ChangeKpiWidgetHeader
    | ChangeKpiWidgetMeasure
    | ChangeKpiWidgetFilterSettings
    | ChangeKpiWidgetComparison
    | RefreshKpiWidget
    | ChangeInsightWidgetHeader
    | ChangeInsightWidgetFilterSettings
    | ChangeInsightWidgetVisProperties
    | ChangeInsightWidgetInsight
    | ModifyDrillsForInsightWidget
    | RemoveDrillsForInsightWidget
    | RefreshInsightWidget
    | CreateAlert
    | UpdateAlert
    | RemoveAlert
    | CreateScheduledEmail
    | Drill
    | DrillDown
    | DrillToAttributeUrl
    | DrillToCustomUrl
    | DrillToDashboard
    | DrillToInsight
    | DrillToLegacyDashboard;
