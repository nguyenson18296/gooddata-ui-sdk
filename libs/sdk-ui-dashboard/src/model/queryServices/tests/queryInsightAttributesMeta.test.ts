// (C) 2021 GoodData Corporation
import { DashboardTester, preloadedTesterFactory } from "../../tests/DashboardTester";
import {
    createTestInsightItem,
    EmptyDashboardIdentifier,
    PivotTableWithRowAndColumnAttributes,
    TestSectionHeader,
    TreemapWithOneMeasureAndViewByDateAndSegmentByDate,
    TreemapWithSingleMeasureAndViewByFilteredToOneElement,
} from "../../tests/Dashboard.fixtures";
import { addLayoutSection } from "../../commands";
import { InsightAttributesMeta, queryInsightAttributesMeta } from "../../queries";
import { insightRef } from "@gooddata/sdk-model";

describe("query insight attributes meta", () => {
    let Tester: DashboardTester;
    beforeEach(preloadedTesterFactory((tester) => (Tester = tester), EmptyDashboardIdentifier));

    it("should return metadata for insight with multiple attributes in buckets", async () => {
        await Tester.dispatchAndWaitFor(
            addLayoutSection(-1, TestSectionHeader, [
                createTestInsightItem(PivotTableWithRowAndColumnAttributes),
            ]),
            "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_ADDED",
        );

        const result: InsightAttributesMeta = await Tester.query(
            queryInsightAttributesMeta(insightRef(PivotTableWithRowAndColumnAttributes)),
        );

        expect(result.displayForms.length).toEqual(4);
        expect(result.attributes.length).toEqual(4);
        expect(result).toMatchSnapshot();
    });

    it("should return metadata for insight with attribute filter", async () => {
        await Tester.dispatchAndWaitFor(
            addLayoutSection(-1, TestSectionHeader, [
                createTestInsightItem(TreemapWithSingleMeasureAndViewByFilteredToOneElement),
            ]),
            "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_ADDED",
        );

        const result: InsightAttributesMeta = await Tester.query(
            queryInsightAttributesMeta(insightRef(TreemapWithSingleMeasureAndViewByFilteredToOneElement)),
        );

        expect(result.displayForms.length).toEqual(1);
        expect(result.attributes.length).toEqual(1);
        expect(result).toMatchSnapshot();
    });

    it("should return metadata for insight with date attributes in buckets", async () => {
        await Tester.dispatchAndWaitFor(
            addLayoutSection(-1, TestSectionHeader, [
                createTestInsightItem(TreemapWithOneMeasureAndViewByDateAndSegmentByDate),
            ]),
            "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_ADDED",
        );

        const result: InsightAttributesMeta = await Tester.query(
            queryInsightAttributesMeta(insightRef(TreemapWithOneMeasureAndViewByDateAndSegmentByDate)),
        );

        expect(result.displayForms.length).toEqual(2);
        expect(result.attributes.length).toEqual(2);
        expect(result).toMatchSnapshot();
    });
});
