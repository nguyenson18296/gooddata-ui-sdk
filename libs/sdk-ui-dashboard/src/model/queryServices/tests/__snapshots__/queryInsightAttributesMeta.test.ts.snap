// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query insight attributes meta should return metadata for insight with attribute filter 1`] = `
Object {
  "attributes": Array [
    Object {
      "deprecated": false,
      "description": "",
      "id": "attr.product.id",
      "production": true,
      "ref": Object {
        "uri": "/gdc/md/referenceworkspace/obj/1054",
      },
      "title": "Product",
      "type": "attribute",
      "unlisted": false,
      "uri": "/gdc/md/referenceworkspace/obj/1054",
    },
  ],
  "displayForms": Array [
    Object {
      "attribute": Object {
        "uri": "/gdc/md/referenceworkspace/obj/1054",
      },
      "deprecated": false,
      "description": "",
      "id": "label.product.id.name",
      "production": true,
      "ref": Object {
        "uri": "/gdc/md/referenceworkspace/obj/1055",
      },
      "title": "Product Name",
      "type": "displayForm",
      "unlisted": false,
      "uri": "/gdc/md/referenceworkspace/obj/1055",
    },
  ],
  "usage": Object {
    "inAttributes": Array [
      Object {
        "identifier": "label.product.id.name",
        "type": "displayForm",
      },
    ],
    "inFilters": Array [
      Object {
        "identifier": "label.product.id.name",
        "type": "displayForm",
      },
    ],
    "inMeasureFilters": Array [],
  },
}
`;

exports[`query insight attributes meta should return metadata for insight with date attributes in buckets 1`] = `
Object {
  "attributes": Array [
    Object {
      "deprecated": false,
      "description": "Year",
      "displayForms": Array [
        Object {
          "attribute": Object {
            "uri": "/gdc/md/referenceworkspace/obj/497",
          },
          "deprecated": false,
          "description": "DisplayForm Year.",
          "id": "closed.aag81lMifn6q",
          "production": true,
          "ref": Object {
            "uri": "/gdc/md/referenceworkspace/obj/515",
          },
          "title": "Year (Closed)",
          "type": "displayForm",
          "unlisted": false,
          "uri": "/gdc/md/referenceworkspace/obj/515",
        },
      ],
      "drillDownStep": Object {
        "uri": "/gdc/md/referenceworkspace/obj/547",
      },
      "id": "closed.year",
      "production": true,
      "ref": Object {
        "uri": "/gdc/md/referenceworkspace/obj/497",
      },
      "title": "Year (Closed)",
      "type": "attribute",
      "unlisted": false,
      "uri": "/gdc/md/referenceworkspace/obj/497",
    },
    Object {
      "deprecated": false,
      "description": "Year",
      "displayForms": Array [
        Object {
          "attribute": Object {
            "uri": "/gdc/md/referenceworkspace/obj/332",
          },
          "deprecated": false,
          "description": "DisplayForm Year.",
          "id": "created.aag81lMifn6q",
          "production": true,
          "ref": Object {
            "uri": "/gdc/md/referenceworkspace/obj/350",
          },
          "title": "Year (Created)",
          "type": "displayForm",
          "unlisted": false,
          "uri": "/gdc/md/referenceworkspace/obj/350",
        },
      ],
      "drillDownStep": Object {
        "uri": "/gdc/md/referenceworkspace/obj/382",
      },
      "id": "created.year",
      "production": true,
      "ref": Object {
        "uri": "/gdc/md/referenceworkspace/obj/332",
      },
      "title": "Year (Created)",
      "type": "attribute",
      "unlisted": false,
      "uri": "/gdc/md/referenceworkspace/obj/332",
    },
  ],
  "displayForms": Array [
    Object {
      "attribute": Object {
        "uri": "/gdc/md/referenceworkspace/obj/497",
      },
      "deprecated": false,
      "description": "DisplayForm Year.",
      "id": "closed.aag81lMifn6q",
      "production": true,
      "ref": Object {
        "uri": "/gdc/md/referenceworkspace/obj/515",
      },
      "title": "Year (Closed)",
      "type": "displayForm",
      "unlisted": false,
      "uri": "/gdc/md/referenceworkspace/obj/515",
    },
    Object {
      "attribute": Object {
        "uri": "/gdc/md/referenceworkspace/obj/332",
      },
      "deprecated": false,
      "description": "DisplayForm Year.",
      "id": "created.aag81lMifn6q",
      "production": true,
      "ref": Object {
        "uri": "/gdc/md/referenceworkspace/obj/350",
      },
      "title": "Year (Created)",
      "type": "displayForm",
      "unlisted": false,
      "uri": "/gdc/md/referenceworkspace/obj/350",
    },
  ],
  "usage": Object {
    "inAttributes": Array [
      Object {
        "identifier": "closed.aag81lMifn6q",
        "type": "displayForm",
      },
      Object {
        "identifier": "created.aag81lMifn6q",
        "type": "displayForm",
      },
    ],
    "inFilters": Array [],
    "inMeasureFilters": Array [],
  },
}
`;

exports[`query insight attributes meta should return metadata for insight with multiple attributes in buckets 1`] = `
Object {
  "attributes": Array [
    Object {
      "deprecated": false,
      "description": "",
      "id": "attr.product.id",
      "production": true,
      "ref": Object {
        "uri": "/gdc/md/referenceworkspace/obj/1054",
      },
      "title": "Product",
      "type": "attribute",
      "unlisted": false,
      "uri": "/gdc/md/referenceworkspace/obj/1054",
    },
    Object {
      "deprecated": false,
      "description": "",
      "id": "attr.owner.department",
      "production": true,
      "ref": Object {
        "uri": "/gdc/md/referenceworkspace/obj/1088",
      },
      "title": "Department",
      "type": "attribute",
      "unlisted": false,
      "uri": "/gdc/md/referenceworkspace/obj/1088",
    },
    Object {
      "deprecated": false,
      "description": "",
      "id": "attr.stage.name",
      "production": true,
      "ref": Object {
        "uri": "/gdc/md/referenceworkspace/obj/1091",
      },
      "title": "Stage Name",
      "type": "attribute",
      "unlisted": false,
      "uri": "/gdc/md/referenceworkspace/obj/1091",
    },
    Object {
      "deprecated": false,
      "description": "",
      "id": "attr.owner.region",
      "production": true,
      "ref": Object {
        "uri": "/gdc/md/referenceworkspace/obj/1086",
      },
      "title": "Region",
      "type": "attribute",
      "unlisted": false,
      "uri": "/gdc/md/referenceworkspace/obj/1086",
    },
  ],
  "displayForms": Array [
    Object {
      "attribute": Object {
        "uri": "/gdc/md/referenceworkspace/obj/1054",
      },
      "deprecated": false,
      "description": "",
      "id": "label.product.id.name",
      "production": true,
      "ref": Object {
        "uri": "/gdc/md/referenceworkspace/obj/1055",
      },
      "title": "Product Name",
      "type": "displayForm",
      "unlisted": false,
      "uri": "/gdc/md/referenceworkspace/obj/1055",
    },
    Object {
      "attribute": Object {
        "uri": "/gdc/md/referenceworkspace/obj/1088",
      },
      "deprecated": false,
      "description": "",
      "id": "label.owner.department",
      "production": true,
      "ref": Object {
        "uri": "/gdc/md/referenceworkspace/obj/1089",
      },
      "title": "Department",
      "type": "displayForm",
      "unlisted": false,
      "uri": "/gdc/md/referenceworkspace/obj/1089",
    },
    Object {
      "attribute": Object {
        "uri": "/gdc/md/referenceworkspace/obj/1091",
      },
      "deprecated": false,
      "description": "",
      "id": "label.stage.name.stagename",
      "production": true,
      "ref": Object {
        "uri": "/gdc/md/referenceworkspace/obj/1092",
      },
      "title": "Stage Name",
      "type": "displayForm",
      "unlisted": false,
      "uri": "/gdc/md/referenceworkspace/obj/1092",
    },
    Object {
      "attribute": Object {
        "uri": "/gdc/md/referenceworkspace/obj/1086",
      },
      "deprecated": false,
      "description": "",
      "id": "label.owner.region",
      "production": true,
      "ref": Object {
        "uri": "/gdc/md/referenceworkspace/obj/1087",
      },
      "title": "Region",
      "type": "displayForm",
      "unlisted": false,
      "uri": "/gdc/md/referenceworkspace/obj/1087",
    },
  ],
  "usage": Object {
    "inAttributes": Array [
      Object {
        "identifier": "label.product.id.name",
        "type": "displayForm",
      },
      Object {
        "identifier": "label.owner.department",
        "type": "displayForm",
      },
      Object {
        "identifier": "label.stage.name.stagename",
        "type": "displayForm",
      },
      Object {
        "identifier": "label.owner.region",
        "type": "displayForm",
      },
    ],
    "inFilters": Array [],
    "inMeasureFilters": Array [],
  },
}
`;
